src/sql/src/plan.rs

```diff
@@ -1298,16 +1298,52 @@ pub struct Ingestion {
 pub struct WebhookValidation {
     /// The expression used to validate a request.
     pub expression: MirScalarExpr,
+    /// Description of the source that will be created.
+    pub relation_desc: RelationDesc,
     /// The column index to provide the request body and whether to provide it as bytes.
     pub bodies: Vec<(usize, bool)>,
     /// The column index to provide the request headers and whether to provide the values as bytes.
-    ///
-    /// TODO(parkmycar): Support filtering down to specific headers.
     pub headers: Vec<(usize, bool)>,
     /// Any secrets that are used in that validation.
     pub secrets: Vec<WebhookValidationSecret>,
 }
 
+impl WebhookValidation {
+    const MAX_REDUCE_TIME: Duration = Duration::from_secs(2);
+
+    /// Attempt to reduce the internal [`MirScalarExpr`] into a simpler expression.
+    ///
+    /// The reduction happens on a separate thread, we also only wait for
+    /// [`WebhookValidation::MAX_REDUCE_TIME`] before timing out and returning an error.
+    pub async fn reduce_expression(&mut self) -> Result<(), &'static str> {
+        let WebhookValidation {
+            expression,
+            relation_desc,
+            ..
+        } = self;
+
+        // On a different thread, attempt to reduce the expression.
+        let mut expression_ = expression.clone();
+        let desc_ = relation_desc.clone();
+        let reduce_task = mz_ore::task::spawn_blocking(
+            || "webhook-validation-reduce",
+            move || {
+                expression_.reduce(&desc_.typ().column_types);
+                expression_
+            },
+        );
+
+        match tokio::time::timeout(Self::MAX_REDUCE_TIME, reduce_task).await {
```

Comment:
I get the idea behind a timeout, but as far as I can tell we don't have similar timeouts on other calls to `reduce`. I would prefer a more holistic approach to timeouts during optimization over ad-hoc timeouts for individual statements. Plus it's generally better for DDL to spend time optimizing upfront so we can save time later every time the validation is run.

Reply:
Also if you move this call to some optimization pass, then we already know how to do optimization off-thread.

Reply:
> Also if you move this call to some optimization pass, then we already know how to do optimization off-thread.

I think currently we do off-thread optimization only for peeks (in `sequence_peek`). (Peeks were the most urgent to move off-thread, because certain user workloads might be doing peeks all the time.)

> as far as I can tell we don't have similar timeouts on other calls to reduce

Yes, `reduce` has a call from `plan_index_exprs`, where it's not off-thread. (Also, `reduce` has a bunch of calls from inside the optimizer (`mz-transform` crate), which are currently off-thread only when we are optimizing a peek.)

I think the long-term plan is that all optimization should happen off-thread. To facilitate this, we'll create some more "official" optimizer interface in a new `mz-optimizer` crate. For `reduce`, the interface could be a wrapper that does some similar task spawning and timeouting as the code in this PR. So having this code for at least this particular `reduce` call seems to be a forward-looking thing: This code might be the basis of what the future official optimizer interface will do in its `reduce` wrapper. (@aalexandrov might also have an opinion here, as he already started creating the official optimizer interface in recent weeks.)

Reply:
Ah ok, in that case I think it's fine as is. Ditto on the sequencing vs planning.

Reply:
Thanks @ggevay, glad this is inline with what that Compute Team is thinking about!

Reply:
I'm not a big fan of the 

```
match tokio::time::timeout(Self::MAX_REDUCE_TIME, reduce_task).await
```

block because it means that this code path is not deterministic. Overall I think forking in a separate thread is fine, but:

1. We should have some (rather large) timeout after which we bail with an error instead of falling back to something.
1. I prefer a commit / PR where we either do this everywhere or nowhere.

As @ggevay pointed out we already call `reduce` on scalars multiple times in the actual optimization path (which is happening on the main coordinator thread), so I would prefer to stick to that approach and create a separate issue to rework this for all optimization paths.

Reply:
Hmm, yes, @aalexandrov is right. I didn't think about the non-determinism issue. If the optimization time for a query is close to the timeout, then it can happen that after a restart (at a materialize version upgrade) we suddenly time out due to random variation, and thus maybe substantially change the performance. And this will be hard to figure out with a fallback at the timeout. Erroring out at least makes it obvious where the problem is.